devito/ops/transformer.py

```diff
@@ -45,17 +61,69 @@ def opsit(trees, count):
         parameters
     )
 
-    stencil_arrays_initializations = itertools.chain(*[
-        to_ops_stencil(p, node_factory.ops_args_accesses[p])
+    # Extract all symbols that need to be converted to ops_dat
+    to_dat = set()
+    symbols = set(FindSymbols('symbolics').visit(trees[0].root))
+    symbols -= set(FindSymbols('defines').visit(trees[0].root))
+    to_dat |= symbols
+
+    # To ensure deterministic code generation we order the datasets to
+    # be generated (since a set is an unordered collection)
+    to_dat = filter_sorted(to_dat)
+
+    name_to_ops_dat = {}
+    pre_loop = []
+    for f in to_dat:
+        if f.is_Constant:
+            continue
+
+        pre_loop.extend(create_ops_dat(f, name_to_ops_dat, block))
+
+    name_to_ops_stencil = {}
+    stencil_arrays_initializations = [
+        to_ops_stencil(p, node_factory.ops_args_accesses[p], name_to_ops_stencil)
         for p in parameters if isinstance(p, OpsAccessible)
-    ])
+    ]
+
+    ops_args = []
+    for p in parameters:
+        ops_args.append(create_ops_arg(p, name_to_ops_dat, name_to_ops_stencil))
+
+    par_loop = create_ops_par_loop(block, ops_kernel, par_loop_array, ops_args)
 
-    pre_time_loop = stencil_arrays_initializations
+    pre_time_loop = itertools.chain(
+        *[pre_loop], *[stencil_arrays_initializations], [par_loop_range])
 
-    return pre_time_loop, ops_kernel
+    return pre_time_loop, par_loop, ops_kernel
+
+
+def create_ops_arg(data, name_to_ops_dat, name_to_ops_stencil):
+    if data.is_Constant:
+        return Call(namespace['ops_arg_gbl'],
+                             [Byref(Constant(name=data.name.replace('*', ''))),
+                              1, Literal('"%s"' % data._C_typedata),
+                              Literal(namespace['ops_read'])])
+    else:
+        return Call(namespace['ops_arg_dat'],
+                             [Literal('%s' % name_to_ops_dat[data.name]), 1,
+                              Literal(name_to_ops_stencil[data.name]),
```

Comment:
Doesn't need to be a Literal

Reply:
Can not, this is a call

Reply:
You can make this a Function

Reply:
I believe a Call is the right thing here.

Reply:
Why?

Reply:
As far as I know
- ops_arg_gbl(T *data, int dim, char *type, ops access acc);
- ops_arg_dat(ops dat dat, ops stencil stencil, char *type, ops access acc);
should mapped to a Devito Call, just as we do with ops_par_loop(...), ops_init(...), ops_exit(). 

I know that in some places Call has not been used, but I would like to keep a standard whenever we **call** something from the OPS API.

This is useful for testing, too.

Reply:
Fair enough. I think however it would prefer to use `Functions` when possible and leave `Call`s for when they are necessary (ops_par_loop, ops_init, ops_exit)

Reply:
> leave Calls for when they can't be avoided

Why do you want to avoid Call?

Reply:
In general I think it's preferable to use other people's stuff when possible. It's one less thing we need to maintain